main/src/library/Pkger/Parser.flix

```diff
@@ -0,0 +1,637 @@
+    /// ==
+    /// Some(('b' :: Nil, Nil))
+    /// ```
+    pub def tailSeperatedAny[a, b](
+        p: Parser[a],
+        sep: Parser[b]
+    ): Parser[List[a]] = tailSeperatedMany(p, sep) `or` pure(Nil)
+
+    /// Given parsers `p` and `sep`, returns a parser which uses `p` one or more times with `sep`
+    /// interlaced. If `sep` can be run after the last run of `p`, the new parser does so. Returns a
+    /// list of values parsed by `p`. Discards anything parsed by `sep`. Fails if `p` couldn't be
+    /// run any times.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// String.toList("ababa") |> tailSeperatedMany(tag('a'), tag('b'))
+    /// ==
+    /// Some((Nil, 'a' :: 'a' :: 'a' :: Nil))
+    /// ```
+    ///
+    /// ```
+    /// String.toList("abab") |> tailSeperatedMany(tag('a'), tag('b'))
+    /// ==
+    /// Some((Nil, 'a' :: 'a' :: Nil))
+    /// ```
+    ///
+    /// ```
+    /// String.toList("a") |> tailSeperatedMany(tag('a'), tag('b'))
+    /// ==
+    /// Some((Nil, 'a' :: Nil))
+    /// ```
+    ///
+    /// ```
+    /// String.toList("b") |> tailSeperatedMany(tag('a'), tag('b'))
+    /// ==
+    /// None
+    /// ```
+    pub def tailSeperatedMany[a, b](
+        p: Parser[a],
+        sep: Parser[b]
+    ): Parser[List[a]] = seperatedMany(p, sep) `skipRight` optional(sep)
+
+    /// Given three parsers `lhs`, `mid`, and `rhs`, returns new parser which starts by applying
+    /// `lhs`, throws away the output, applies `mid`, saves the output, and applies `rhs` and
+    /// returns the output from mid.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// String.toList("[abz]") |>
+    ///     delimit(tag('['), takeAny(char().filter(Char.isLetter)), tag(']'))
+    /// ==
+    /// Some((Nil, 'a' :: 'b' :: 'z' :: Nil))
+    /// ```
+    ///
+    /// ```
+    /// String.toList("[a.bz]") |>
+    ///     delimit(tag('['), takeAny(char().filter(Char.isLetter)), tag(']'))
+    /// ==
+    /// None
+    /// ```
+    ///
+    /// ```
+    /// String.toList("[]") |>
+    ///     delimit(tag('['), takeAny(char().filter(Char.isLetter)), tag(']'))
+    /// ==
+    /// Some((Nil, Nil))
+    /// ```
+    pub def delimit[a, b, c](
+        lhs: Parser[a],
+        mid: Parser[b],
+        rhs: Parser[c]
+    ): Parser[b] = lhs `skipLeft` mid `skipRight` rhs
+
+    /// Given a Parser `p`, an initial value `initial`, and a function `f`, generates a parser,
+    /// which recursively applies `f` to initial, or the last output from `f`, and the next parsed
+    /// value.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// String.toList("abc") |> fold(char(), Nil, l -> c -> c :: l)
+    /// ==
+    /// Some((Nil, 'c' :: 'b' :: 'a' :: Nil))
+    /// ```
+    pub def fold[a, b](
+        p: Parser[a],
+        initial: b,
+        f: b -> a -> b
+    ): Parser[b] = input -> Some(foldInternal(p, input, initial, f))
+
+    //! Helper function for fold
+    def foldInternal[a, b](
```

Comment:
We usually call these `helper`, i.e. `foldHelper`. The name is not great, but it has already become convention.